### ISSUE TITLE
Prefix / Starts With filtering support in filter_by

### ISSUE THREAD

**Issue by brettwillis** [typesense/typesense#696]
The query parameter `q` and filter parameter `filter_by` support prefix matching by default, and if enabled in the schema, the `q` parameter supports infix. However, even if `infix` is enabled in the schema:
- `filter_by` on a `string` field **does not support infix** (it does support prefix)
- `filter_by` on an array `string[]` field **does not support prefix or infix**

### How to reproduce

Consider the schema:

```JSON
{
  "fields": [
    {
      "name": "label",
      "type": "string",
      "infix": true
    },
    {
      "name": "tags",
      "type": "string[]",
      "infix": true
    }
  ]
}
```

And the document:

```JSON
{
  "label": "abcd",
  "tags": ["season1/episode1", "season2/episode2"]
}
```

| Match type                            | Query                                              | Expected match    | Actual match    |
|---------------------------------|--------------------------------------|--------------------|------------------|
| `q` prefix on `string`               | `?q=abc&query_by=label`             | Yes                        | ✅ Yes              |
| `q` infix on `string`                 | `?q=bcd&query_by=label`             | Yes                        | ✅ Yes              |
| `q` prefix on `string[]`             | `?q=season1&query_by=tags`      | Yes                        | ✅ Yes              |
| `q` infix on `string[]`                | `?q=episode1&query_by=tags`    | Yes                        | ✅ Yes              |
| `filter_by` prefix on `string`     | `?q=&query_by=label&filter_by=label:abc`            | Yes                        | ✅ Yes              |
| `filter_by` infix on `string`       | `?q=&query_by=label&filter_by=label:bcd`            | Yes                        |  ❌ No match              |
| `filter_by` prefix on `string[]`  | `?q=&query_by=label&filter_by=tags:season1`        | Yes                        | ❌ No match              |
| `filter_by` infix on `string[]`    | `?q=&query_by=label&filter_by=tags:episode1`       | Yes                        | ❌ No match              |

Typsense Version: `0.23.1`
OS: Linux

**Comment by kishorenc** [comment]
@brettwillis 

Thank you for creating this amazingly detailed issue.

By design, `filter_by` does not work on prefix/infix searches. You _could_ prefix search on the facet values returned via the `facet_query` parameter, but otherwise you can't do it using the `filter_by` clause.

I'm surprised to see that `filter_by` prefix on `string` working, but I'm unable to reproduce it, see this gist: https://gist.github.com/kishorenc/957bbf1aadee622aa71d81f78c1fd147

**Comment by brettwillis** [comment]
@kishorenc you're right, `filter_by` doesn't support prefix match either. What I was seeing as "prefix match" was actually matching a complete word/token, because there was a token word/separator (I was using different data).

Is it on the to-do list for `filter_by` to support prefix and infix? The presense of an inexact match operator `:` and an exact match operator `:=` initially suggested to me that `filter_by:...` (inexact) may support the same prefix/infix semantics as queries. Whereas if someone wanted to improve performance, they could use the exact match`:=` operator. Is there a disadvantage or difficulty to support prefix/infix?

It would be useful for me, otherwise I currently need to duplicate a lot of data to split out all the possible token combinations. Consider the following queries from the trivial example:
- Search within those documents that have a "season1": `?q=...&query_by=label&filter_by=tags:season1` ❌  no prefix filter
- Search within those documents that have an "episode2": `?q=...&query_by=label&filter_by=tags:episode2` ❌  no infix filter
- Search within those documents that have a "season1/episode1": `?q=...&query_by=label&filter_by=tags:season1/episode1` ✅ exact match

**Comment by brettwillis** [comment]
Understood. It would be great to add to your backlog, thank you! In the meantime, I work around this by duplicating/splitting out into all the different tokens required to support my queries.

> For prefix search we might need a new operator like maybe filter_by=tag:run*.

Hmm in my opinion, wildcard matching could be a completely different feature. E.g. it could also be `tag:=*shoes` (prefix), `tag:=run*shoes` (infix), `color:=gr?y` etc. which would apply to `q` and `facet_query` also. Whereas what we're talking about here is just behaviour parity between `filter_by` and `q`/`facet_query`, so I would personally be comfortable keeping the same `:` operator (just like `facet_query`, except we are also talking about infix matching not just prefix).

**Comment by manasomali** [comment]
I would also like the feature of prefix (or infix) queries to filter by clause.
In my case, I have a main field to be used as the search term and another two (or more) fields to be used as filters. The request is to get the documents that match the filters even if it's not an exact match (e.g. substring match).

I believe I can work around this by using multi-search, performing the search in each filter field step by step, and joining the hits results.
Once the feature is implemented, I will update the implementation.

Congratulations on the project and organization! 👍🏻

**Comment by devid-verraes** [comment]
We are also highly interested in this feature. Our application has a general search field which you can combine with additional filters to limit the results even more. Most of these additional filters work fine with an exact match, but some of them require a prefix search. For example  a use case we have is when you want to look for all the invoices for a customer where the payment reference start with a given text.

**Comment by manasomali** [comment]
Update, I could not implement it with multi-search 😢.
The workaround will be duplicating/splitting the filter field (@brettwillis).

The idea is:
Instead of index only:
- saitama.caped.baldy@mail.com

Will be indexed:
- saitama.caped.baldy@mail.com
- saitama.caped
- caped.baldy
- baldy@mail.com
- saitama
- caped
- baldy
(maybe more or less variations, depends on the problem and data type)

It has limitations, but it is the best workaround for it (I believe).

**Comment by kishorenc** [comment]
Thanks for all the feedback. This is definitely on our medium term roadmap, but we are tied up with other features at the moment. We hope hope to pick this up in a couple of months.

**Comment by mindblight** [comment]
I would also love if this were available. Another really valuable use case: Combining prefix matching w/ scoped API keys could unlock really powerful authorization patterns. 

E.g. Lets say I have multiple files from a file system:
```
- dir1/
  - dir2/
    - file1
    - file2
  - dir3/
    - file3
```

I could put the `path` into each record and use scoped API keys to allow a user to only search `dir1/dir2`. More complex URI schemes could be useful for search in multi-tenent apps too.

**Comment by jasonbosco** [comment]
Prefix filtering is now available in `0.26.0.rc62`. You want to end the filter string with a `*` to enable this.

Eg:

`filter_by: (company_name:Acm* && country:Un*) || city: Los*`

This will return records where (one of the words in) `company_name` starts with `Acm` and (one of the words in) `country` starts with `Un` or records where (one of the words in) `city` starts with `Los`

**Comment by jasonbosco** [comment]
Released in v26
